packages/data/src/components/use-select/index.js

```diff
@@ -238,3 +238,128 @@ export default function useSelect( mapSelect, deps ) {
+
+	const latestRegistry = useRef( registry );
+	const latestMapSelect = useRef();
+	const latestIsAsync = useRef( isAsync );
+	const latestMapOutput = useRef();
+	const latestMapOutputError = useRef();
+
+	// Keep track of the stores being selected in the `mapSelect` function,
+	// and only subscribe to those stores later.
+	const listeningStores = useRef( [] );
+	const wrapSelect = useCallback(
+		( callback ) =>
+			registry.__unstableMarkListeningStores(
+				() => callback( registry.suspendSelect, registry ),
+				listeningStores
+			),
+		[ registry ]
+	);
+
+	let mapOutput = latestMapOutput.current;
+	let mapOutputError = latestMapOutputError.current;
+
+	const hasReplacedRegistry = latestRegistry.current !== registry;
+	const hasReplacedMapSelect = latestMapSelect.current !== _mapSelect;
+
+	if ( hasReplacedRegistry || hasReplacedMapSelect ) {
+		try {
+			mapOutput = wrapSelect( _mapSelect );
+		} catch ( error ) {
+			mapOutputError = error;
+		}
+	}
+
+	useIsomorphicLayoutEffect( () => {
+		latestRegistry.current = registry;
+		latestMapSelect.current = _mapSelect;
+		latestMapOutput.current = mapOutput;
+		latestMapOutputError.current = mapOutputError;
+
+		// This has to run after the other ref updates
+		// to avoid using stale values in the flushed
+		// callbacks or potentially overwriting a
+		// changed `latestMapOutput.current`.
+		if ( latestIsAsync.current !== isAsync ) {
+			latestIsAsync.current = isAsync;
+			renderQueue.flush( queueContext );
+		}
+	} );
+
+	// Generate a "flag" for used in the effect dependency array.
+	// It's different than just using `mapSelect` since deps could be undefined,
+	// in that case, we would still want to memoize it.
+	const depsChangedFlag = useMemo( () => ( {} ), deps || [] );
+
+	useIsomorphicLayoutEffect( () => {
+		const onStoreChange = () => {
+			try {
+				const newMapOutput = wrapSelect( latestMapSelect.current );
+
+				if ( isShallowEqual( latestMapOutput.current, newMapOutput ) ) {
+					return;
+				}
+				latestMapOutput.current = newMapOutput;
+			} catch ( error ) {
+				latestMapOutputError.current = error;
+			}
+
+			forceRender();
+		};
+
+		const onChange = () => {
+			if ( latestIsAsync.current ) {
+				renderQueue.add( queueContext, onStoreChange );
+			} else {
+				onStoreChange();
+			}
+		};
+
+		// catch any possible state changes during mount before the subscription
+		// could be set.
+		onStoreChange();
+
+		const unsubscribers = listeningStores.current.map( ( storeName ) =>
+			registry.__unstableSubscribeStore( storeName, onChange )
+		);
+
+		return () => {
+			// The return value of the subscribe function could be undefined if the store is a custom generic store.
+			unsubscribers.forEach( ( unsubscribe ) => unsubscribe?.() );
+			renderQueue.cancel( queueContext );
+		};
+	}, [ registry, wrapSelect, depsChangedFlag ] );
+
+	if ( mapOutputError ) {
+		throw mapOutputError;
+	}
+
+	return mapOutput;
+}
```

Comment:
How different is the code here from the "raw" useSelect? Asking to see if we can share the same hook (or at least same base implementation) or if it's completely different?

Reply:
Nevermind, I just saw your comment above :P

Reply:
The biggest difference, not easy to wrap in some `if ( suspenseMode )` conditions, is how errors thrown by `mapSelect` are handled. `useSelect` kind of doesn't know what to do with them, it logs them to console, and otherwise tries to ignore them: it returns the previous (successful) `mapSelect` result, and hopes maybe on the next call the error will go away.

`useSuspenseSelect` simply rethrows the error, and lets suspense or error boundary to catch it.